packages/myst-execute/src/execute.ts

```diff
@@ -0,0 +1,347 @@
+import { select, selectAll } from 'unist-util-select';
+import type { PageFrontmatter } from 'myst-frontmatter';
+import type { Kernel, KernelMessage, Session, SessionManager } from '@jupyterlab/services';
+import type { Code, InlineExpression } from 'myst-spec-ext';
+import type { IOutput } from '@jupyterlab/nbformat';
+import type { GenericNode, GenericParent, IExpressionResult } from 'myst-common';
+import { fileError, fileInfo, fileWarn } from 'myst-common';
+import type { VFile } from 'vfile';
+import path from 'node:path';
+import assert from 'node:assert';
+import { createHash } from 'node:crypto';
+import type { Plugin } from 'unified';
+import type { ICache } from './cache.js';
+
+/**
+ * Interpret an IOPub message as an IOutput object
+ *
+ * @param msg IOPub message
+ */
+function IOPubAsOutput(msg: KernelMessage.IIOPubMessage): IOutput {
+  return {
+    output_type: msg.header.msg_type,
+    ...msg.content,
+  };
+}
+
+/**
+ * Execute a code string in the given kernel, returning any outputs.
+ *
+ * @param kernel connection to an active kernel
+ * @param code code to execute
+ */
+async function executeCode(kernel: Kernel.IKernelConnection, code: string) {
+  const future = kernel.requestExecute({
+    code: code,
+  });
+
+  const outputs: IOutput[] = [];
+  future.onIOPub = (msg) => {
+    // Only listen for replies to this execution
+    if (
+      (msg.parent_header as any).msg_id !== future.msg.header.msg_id ||
+      (msg.parent_header as any).msg_type !== 'execute_request'
+    ) {
+      console.debug('Ignoring IOPub reply');
```

Comment:
Can we use the `session.log.debug` here? Or at the least pass in the `log` where we are piping things?